sdks/java/io/datalake/src/main/java/org/apache/beam/sdk/io/datalake/DataLakeIO.java

```diff
@@ -0,0 +1,530 @@
+                } catch (NoSuchSchemaException e) {
+                    LOG.warn(
+                            "Unable to infer a schema for type {}. Attempting to infer a coder without a schema.",
+                            outputType);
+                }
+                try {
+                    return registry.getCoder(outputType);
+                } catch (CannotProvideCoderException e) {
+                    LOG.warn("Unable to infer a coder for type {}", outputType);
+                    return null;
+                }
+            }
+        }
+
+        @Override
+        public PCollection<OutputT> expand(PCollection<ParameterT> input) {
+            Coder<OutputT> coder =
+                    inferCoder(
+                            input.getPipeline().getCoderRegistry(), input.getPipeline().getSchemaRegistry());
+            checkNotNull(
+                    coder,
+                    "Unable to infer a coder for DataLakeIO.readAll() transform. "
+                            + "Provide a coder via withCoder, or ensure that one can be inferred from the"
+                            + " provided RowMapper.");
+            PCollection<OutputT> output =
+                    input
+                            .apply(
+                                    ParDo.of(
+                                            new DataLakeIO.ReadFn<>(
+                                                    getPath(),
+                                                    getFormat(),
+                                                    getSparkConf(),
+                                                    getRowMapper()
+                                            )))
+                            .setCoder(coder);
+
+            try {
+                TypeDescriptor<OutputT> typeDesc = coder.getEncodedTypeDescriptor();
+                SchemaRegistry registry = input.getPipeline().getSchemaRegistry();
+                Schema schema = registry.getSchema(typeDesc);
+                output.setSchema(
+                        schema,
+                        typeDesc,
+                        registry.getToRowFunction(typeDesc),
+                        registry.getFromRowFunction(typeDesc));
+            } catch (NoSuchSchemaException e) {
+                // ignore
+            }
+
+            return output;
+        }
+
+        @Override
+        public void populateDisplayData(DisplayData.Builder builder) {
+            super.populateDisplayData(builder);
+            if(getQuery() != null){
+                builder.add(DisplayData.item("query", getQuery()));
+            }
+            builder.add(DisplayData.item("rowMapper", getRowMapper().getClass().getName()));
+            if (getCoder() != null) {
+                builder.add(DisplayData.item("coder", getCoder().getClass().getName()));
+            }
+        }
+    }
+
+    private static class ReadFn<ParameterT, OutputT> extends DoFn<ParameterT, OutputT> {
```

Comment:
Currently, we are trying to have all sources implemented using the SplittableDoFn pattern to enable scalability, and are doing our best to not include new sources that are not implemented as SDFs. Can this be re-implemented as an SDF instead?

Reply:
Thanks, I'll learn SplittableDoFn, then re-implemente it  as an SDF instead.